File: Ruby/BucketSort/list_pointer_bucket_sort.rb
require 'benchmark'
require 'objspace'

class Rating
  attr_accessor :user_id, :movie_id, :rating, :timestamp
  def initialize(user_id, movie_id, rating, timestamp)
    @user_id = user_id
    @movie_id = movie_id
    @rating = rating
    @timestamp = timestamp
  end

  def to_s
    "#{@user_id},#{@movie_id},#{@rating},#{@timestamp}"
  end
end

class Node
  attr_accessor :rating, :next, :prev
  def initialize(rating)
    @rating = rating
    @next = nil
    @prev = nil
  end
end

class ListPointerBucketSort
  attr_accessor :head, :tail, :size

  def initialize
    @head = @tail = nil
    @size = 0
  end

  def add(rating)
    node = Node.new(rating)
    if @head.nil?
      @head = @tail = node
    else
      @tail.next = node
      node.prev = @tail
      @tail = node
    end
    @size += 1
  end

  def bucket_sort
    buckets = Array.new(11) { [] }
    cur = @head
    while cur
      idx = [(cur.rating.rating * 2).floor, 10].min  # Garante índice válido
      buckets[idx] << cur
      cur = cur.next
    end

    @head = @tail = nil
    @size = 0

    buckets.each do |bucket|
      bucket.each do |node|
        node.next = node.prev = nil
        if @head.nil?
          @head = @tail = node
        else
          @tail.next = node
          node.prev = @tail
          @tail = node
        end
        @size += 1
      end
    end
  end
end

# ----------- INÍCIO DA EXECUÇÃO -------------

if ARGV.size < 2
  puts "Uso: ruby #{__FILE__} <arquivo.csv> <N>"
  exit
end

path = ARGV[0]
n = ARGV[1].to_i
lista = ListPointerBucketSort.new

File.open(path) do |f|
  f.readline # Pula o cabeçalho
  f.each_line.take(n).each do |line|
    user_id, movie_id, rating, timestamp = line.strip.split(',')
    lista.add(Rating.new(user_id.to_i, movie_id.to_i, rating.to_f, timestamp.to_i))
  end
end

mem_before = ObjectSpace.memsize_of_all
time = Benchmark.realtime { lista.bucket_sort }
mem_after = ObjectSpace.memsize_of_all

puts "Ordenado #{n} registros em #{(time * 1000).round(2)} ms"
puts "Memória usada: #{((mem_after - mem_before) / 1024.0).round(2)} KB"

puts "\nPrimeiros 100 registros ordenados:"
cur = lista.head
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.next
end

puts "\nÚltimos 100 registros ordenados:"
cur = lista.tail
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.prev
end

puts "\nTamanho da lista: #{lista.size}"

File: JAVA/Bucket Sort/ListPointerBucketSort.java
import java.io.*;
import java.util.ArrayList;

public class ListPointerBucketSort
{
    static class Rating
    {
        int u,m;float r;long t;
        Rating(int u,int m,float r,long t){this.u=u;this.m=m;this.r=r;this.t=t;}
        public String toString(){return u+","+m+","+r+","+t;}
    }

    static class Node{Rating val;Node nxt,prv;Node(Rating v){val=v;}}

    Node head,tail; int size=0;

    void add(Rating r)
    {
        Node n=new Node(r);if(head==null)head=tail=n;
        else{tail.nxt=n;n.prv=tail;tail=n;}size++;
    }

    void bucketSort()
    {
        @SuppressWarnings("unchecked")
        ArrayList<Node>[] buckets=new ArrayList[11];
        for(int i=0;i<buckets.length;i++) buckets[i]=new ArrayList<>();

        for(Node p=head;p!=null;p=p.nxt)
        {
            int idx=(int)Math.floor(p.val.r*2);
            buckets[idx].add(p);
        }

        head=tail=null; size=0;

        for(ArrayList<Node>b:buckets)
            for(Node n:b)
            {
                n.nxt=n.prv=null;
                if(head==null)head=tail=n;else{tail.nxt=n;n.prv=tail;tail=n;}
                size++;
            }
    }

    public static void main(String[]a)
    throws Exception
    {
        if(a.length<2)
        {
            System.out.println("Uso: java ListPointerBucketSort <ratings.csv> <N>");
            return;
        }

        int N=Integer.parseInt(a[1]); ListPointerBucketSort L=new ListPointerBucketSort();

        try(BufferedReader br=new BufferedReader(new FileReader(a[0])))
        {
            String line=br.readLine();int c=0;
            while((line=br.readLine())!=null&&c<N)
            {
                String[] f=line.split(",");
                L.add(new Rating(Integer.parseInt(f[0]),Integer.parseInt(f[1]),
                                 Float.parseFloat(f[2]),Long.parseLong(f[3])));
                c++;
            }
        }
        
        long t0=System.nanoTime();L.bucketSort();long t1=System.nanoTime();
        System.out.printf("Bucket sort lista ponteiro %d regs: %.2f ms%n",N,(t1-t0)/1e6);
        Node p=L.head;for(int i=0;i<20&&p!=null;i++,p=p.nxt)System.out.println(p.val);
    }
}
